### ISSUE TITLE
Minimizer - API and specs

### ISSUE THREAD

**Issue by mayou36** [zfit/zfit#13]
# Documentation of API

## crucial design options

- [ ] instantiate with the function like `minimizer = Minimizer(func=loss,...)` (RooFit, TFScipy,... -like). (Alternative: pass it to `minimize`)

Base class:  
methods for minimization:

- `__init__(loss, minimizer_options)`
- `minimize(params)` -> minimizes the nll fully (like MINUIT, contrary to TF)
(- for stepwise minimizers, add a "step()" method, to perform one step (calc grad, update vars), but not in the base class)
Attributes:
- parameters -> getter/setters to access the variables
  - `get_parameters(floating=True)`
  - `set_parameters(params)`
- `edm` -> Estimated Distance to Minimum (property)
- `fmin` -> minimum of function (property)
(optional: - state -> getter/setter for state to retrieve the optimizer state)
- tolerance -> tolerance, how close to the minimum to stop
- get_state(copy=True)
- MinimizerState -> some kind of FitResult (also returned by `minimize`)

For the error calculation, methods:
- `hesse(params, sess)` (or similar) -> Hessian of the current state
- `_hesse()` should return a dict_like object with {par1_name: {'error': num, ...}, par2_name:...}
- `error(params, sess)` -> some MC or similar way to determine error (like minos)
- `_error()` should return a dict_like object with {par1_name: {'lower_error': num,  'upper_error',...}, par2_name:...}
Attributes:
- (error -> getter/setter (priors!) for the errors)
(-probably more for the different kind of errors, like hessian_error, mc_error etc.)

 - set_error_method(str_or_func) -> sets what `error` does, which func to use
 - set_error_options(sigma=1, ...) -> options for the error calculation

**Comment by apuignav** [comment]
I think right now we want to keep ROOT/MINUIT as mandatory, at least until a better alternative is found and studied. If this means we don't test certain things, I'm fine with this for the time being.

However, I would like to perform a refactoring to decouple MINUIT from `core` so we just provide an API and MINUIT can be easily replaced afterwards. Sounds like a plan?

**Comment by mayou36** [comment]
Yes, that sounds reasonable, fully agree!

**Comment by apuignav** [comment]
How would the API go then? 

Should we just use something very similar to MINUIT or do we need something more elaborate?

**Comment by mayou36** [comment]
I think too elaborate does not make sense. Looking around with other minimizers (tf, keras, pytorch, scipy...) it seems that there are some commonalities around, "similar" to MINUIT. And nothing very fancy is around.

So to share the idea, I would propose to have the following:

Minimal example:
```
opt = MyOptimizer(config goes here)
var_fmin = opt.minimize(nll, my_vars)
...
```
An optimizer is expected with the following API:
(for common minimizers, like MINUIT or Tensorflow optimizers, we may just write the wrapper so that it is easy to invoke them).

Base class:  
methods for minimization:
- minimize(nll, var, n_calls?) -> minimizes the nll fully (like MINUIT, contrary to TF)
(- for stepwise minimizers, add a "step()" method, to perform one step (calc grad, update vars), but not in the base class)
Attributes:
- variables -> getter/setters to access the variables
- edm -> Estimated Distance to Minimum
- fmin -> minimum of function
(optional: - state -> getter/setter for state to retrieve the optimizer state)
- tol -> tolerance, how close to the minimum to stop

For the error calculation, methods:
- hesse() (or similar) -> Hessian of the current state
- mc_error(var, sigma) -> some MC or similar way to determine error (like minos)
Attributes:
- error -> getter/setter (priors!) for the errors
(-probably more for the different kind of errors, like hessian_error, mc_error etc.)

What's your opinion on that, @apuignav ?

**Comment by rsilvaco** [comment]
Just one minor comment on the error calculation. I think it is inevitable that we need to provide/use Feldman Cousin to calculate the uncertainties, even in the current amplitude analysis that we are doing. Therefore, this should be also another attribute of the minimiser. I'm actually interest to see how efficiently this could be implemented, since it requires a reasonable large number of toys.

**Comment by apuignav** [comment]
Mmm, I think Feldman Cousins is actually independent on the fitter, isn't it? This means that we can implement it regardless of the fitter, and then the API we're discussing guarantees that it will always work.

**Comment by apuignav** [comment]
I have a couple of comments:

  + Regarding `mc_error(var, sigma) -> some MC or similar way to determine error (like minos)`, Minos is not really doing MC, but rather scanning and refitting, so maybe we need another name (and also leave room for minimiser-specific error functions implemented in a generic way).
 + We need profile likelihood calculation, should it go in the minimiser or as an "external" helper? @rsilvaco what do you think? I guess `profile = zfit.minimize.profile_likelihood(nll, [param_list])`      could be enough, and no need to integrate the profile likelihood in the minimiser (similar to Feldman-Cousins).

**Comment by mayou36** [comment]
* true, just the contrast to analytical
 * should we create a `minimizer.error` namespace? And put any error calcs in there? So that the minimizer can have as many methods as it wants? But not my favorite approach...
 * agree to the `profile_nll`, I think it should not stick to the minimizer

**Comment by apuignav** [comment]
There's a final decision to be made: what does `minimize` return? Or, in other words, how does the user get the parameters of the fit? Through repeated calls to the `minimizer` object? For example, RooFit returns a RooFitResult object, which is very practical when you want to repeat the fit many times. In this case we're taking a different approach (you would have to create a new minimizer every time). @rsilvaco @mayou36 Thoughts?

**Comment by apuignav** [comment]
Let's do that. @rsilvaco agrees.

**Comment by apuignav** [comment]
So, discussing it we arrived at the following points. The minimization methods are:

  - `__init__` should not get the loss function. The loss should be passed to the `minimize` function.
  - `minimize` gets loss and `var_list`. No extra arguments. Nothing else should be passed. It returns a `FitResult` (per #37). Maybe, `FitResult` should be called `MinimizerState` (see next)?
  - `hesse` (or something similar), but this uses the current state so maybe it can get a `FitResult` as optional parameter. Otherwise it uses the current state.
  - `error` would correspond to the Minos situation.
  - Getters/setters for variables, tolerance, etc. And EDM, fmin, etc could be accessed through the `MinimizerState`. This is a bit cumbersome, @mayou36 how would you deal with this?

Things like Feldman-Cousins and other methods will go in the `zfit-stats` repo.

**Comment by apuignav** [comment]
Update: the loss function goes into the `__init__`.

No `FitResult` is returned in `minimize`. We need a `get_state` function to get the `MinimizerState`, and a `reset_state` that gets a `MinimizerState` and loads it.

`error` would implement the default error calculation and would get an optional `var_list` argument. We could have a `set_error_method` and `set_error_options` so users can customise this.

**Comment by mayou36** [comment]
> No FitResult is returned in minimize

What do we want to return? I'd propose to return the current state actually. Or what else? (I mean it's also "for free" to return something else then None, right?)

**Comment by mayou36** [comment]
And on names for the method parameters: `var_list`? As we otherwise call them `FitParameters`, I'd propose `parameters` or ("better") `params`. Most of all: consistent, right? @apuignav

**Comment by mayou36** [comment]
should `error` take no arguments at all? What about `params`? To say for which parameters we want to have the error? If None, then take all.

**Comment by mayou36** [comment]
To get the state, should we have a `get_minimzer_state(copy=True)` method? `copy=True` is the behavior you would expect from a fixed "FitResult", while `False` basically gives a "view" (no idea why to use it currently, but why not?)

**Comment by apuignav** [comment]
> > No FitResult is returned in minimize
> 
> What do we want to return? I'd propose to return the current state actually. Or what else? (I mean it's also "for free" to return something else then None, right?)

I would say we want to return the minimizer status.

**Comment by apuignav** [comment]
> And on names for the method parameters: `var_list`? As we otherwise call them `FitParameters`, I'd propose `parameters` or ("better") `params`. Most of all: consistent, right? @apuignav

`params` is good.

**Comment by apuignav** [comment]
> should `error` take no arguments at all? What about `params`? To say for which parameters we want to have the error? If None, then take all.

Exactly, error should take `params=None` and if `None`, all parameters are taken.

**Comment by apuignav** [comment]
> To get the state, should we have a `get_minimzer_state(copy=True)` method? `copy=True` is the behavior you would expect from a fixed "FitResult", while `False` basically gives a "view" (no idea why to use it currently, but why not?). Or always a copy?
> 
> It's not so dangerous I think, as it is a "view-only" object (setters are private, ok? It's only to store and update after a fit only, right?) @apuignav thoughts?

OK, why not? It may prove useful!

**Comment by mayou36** [comment]
Agree with everything!

**Comment by mayou36** [comment]
Yes. (except of maybe additional **kwargs, but we'll see).

**Comment by mayou36** [comment]
One thing missing: the sigma? So the error-definition? should we set that in the minimizer? Like errordef?

**Comment by apuignav** [comment]
Mmm, the sigma is a property of the loss function, isn't it? @rsilvaco?

**Comment by mayou36** [comment]
I think so, yes. But you somehow have to specify how much the loss function should deviate to define the error on the variables. So whatever loss you have, there needs to be an errordef, right? The question may remains to what the value *exactly* means?

**Comment by apuignav** [comment]
I mean, the errordef is not related to the minimizer, it's what 1 sigma (or whatever) means with the loss function you have in there. So it's not a property of the minimizer, it should go to the nll I think.

**Comment by mayou36** [comment]
Agree! Only question? how should they "communicate"?

**Comment by apuignav** [comment]
This is a good question. The only evident solution is to wrap the loss function in a class that contains the errordef...

**Comment by mayou36** [comment]
Yes... didn't wanna say it out loud.

Anyway, there are two thing: define what 1 sigma is in term of loss (-> belongs to the loss) and define for how many sigmas to calculate the error (of the params) for (-> belongs to the `error()` method, right?). We could say this can be set with the `error_options` but it's such an "important" option, I would propose to use a "sigma" or similar argument. (`error_options` to me is more like "additional error algorithm specific options, agree?)

**Comment by apuignav** [comment]
Not sure TBH. I don't know how well the concept of more than 1 sigma is defined in things like Minos, for example.

Even the nomenclature is wrong, I think: sigma is not a good name (@rsilvaco?). I think error should return the upper and lower limits according to the error def in the loss "class".

**Comment by mayou36** [comment]
Issue moved to [zfit/zfit-developement #7](https://github.com/zfit/zfit-developement/issues/7) via [**ZenHub**](https://www.zenhub.com/)
